extensions/systemd/ConsumeJournald.cpp

```diff
@@ -0,0 +1,262 @@
+}
+
+void ConsumeJournald::notifyStop() {
+  bool running = true;
+  if (!running_.compare_exchange_strong(running, false, std::memory_order_acq_rel) || !journal_) return;
+  worker_->enqueue([this] {
+    journal_ = nullptr;
+  }).get();
+  worker_ = nullptr;
+}
+
+void ConsumeJournald::onSchedule(core::ProcessContext* const context, core::ProcessSessionFactory* const sessionFactory) {
+  gsl_Expects(context && sessionFactory && !running_ && worker_);
+  using JournalTypeEnum = systemd::JournalType;
+
+  const auto parse_payload_format = [](const std::string& property_value) -> utils::optional<systemd::PayloadFormat> {
+    if (utils::StringUtils::equalsIgnoreCase(property_value, PAYLOAD_FORMAT_RAW)) return systemd::PayloadFormat::Raw;
+    if (utils::StringUtils::equalsIgnoreCase(property_value, PAYLOAD_FORMAT_SYSLOG)) return systemd::PayloadFormat::Syslog;
+    return utils::nullopt;
+  };
+  const auto parse_journal_type = [](const std::string& property_value) -> utils::optional<JournalTypeEnum> {
+    if (utils::StringUtils::equalsIgnoreCase(property_value, JOURNAL_TYPE_USER)) return JournalTypeEnum::User;
+    if (utils::StringUtils::equalsIgnoreCase(property_value, JOURNAL_TYPE_SYSTEM)) return JournalTypeEnum::System;
+    if (utils::StringUtils::equalsIgnoreCase(property_value, JOURNAL_TYPE_BOTH)) return JournalTypeEnum::Both;
+    return utils::nullopt;
+  };
+  batch_size_ = context->getProperty<size_t>(BatchSize).value_or(10);
+  payload_format_ = (context->getProperty(PayloadFormat) | utils::flatMap(parse_payload_format)).value_or(systemd::PayloadFormat::Syslog);
+  include_timestamp_ = context->getProperty<bool>(IncludeTimestamp).value_or(true);
+  const auto journal_type = (context->getProperty(JournalType) | utils::flatMap(parse_journal_type)).value_or(JournalTypeEnum::System);
+  const auto process_old_messages = context->getProperty<bool>(ProcessOldMessages).value_or(false);
+  timestamp_format_ = [&context] {
+    auto tf_prop = context->getProperty(TimestampFormat).value_or(TimestampFormat.getDefaultValue());
+    if (tf_prop == "ISO" || tf_prop == "ISO 8601" || tf_prop == "ISO8601") return std::string{"%FT%T%Ez"};
+    return tf_prop;
+  }();
+
+  state_manager_ = context->getStateManager();
+  // All journal-related API calls are thread-agnostic, meaning they need to be called from the same thread. In our environment,
+  // where a processor can easily be scheduled on different threads, we ensure this by executing all library calls on a dedicated
+  // worker thread. This is why all such operations are dispatched to a thread and immediately waited for in the initiating thread.
+  journal_ = worker_->enqueue([this, journal_type]{ return libwrapper_->openJournal(journal_type); }).get();
+  const auto seek_default = [this, process_old_messages](libwrapper::Journal& journal) {
+    if (process_old_messages) journal.seekHead();
+    else journal.seekTail();
+    state_manager_->set({{"cursor", getCursor()}});
```

Comment:
When I tried to run the processor locally (both on Ubuntu 18.04 and manjaro) the onSchedule failed because ​this getCursor calls returned an error.
sd_journal_get_cursor: Cannot assign requested address

From the man I think we are missing a sd_journal_next call (I hacked it into here, and now it seems working)

From the man
```
Note that sd_journal_get_cursor() will not work before
      ​sd_journal_next(3) (or related call) has been called at least
      ​once, in order to position the read pointer at a valid entry.

``

Reply:
Removed this call in f936b60 and updated the test to not expect state before a flow file is produced. Now the initial state is only stored after the first flow file is produced. This should not be a problem, but is a slight behavior change.